File: haskell/aoc2015/src/Day04.hs
module Day04
  ( parts
  ) where

import Data.ByteString.Lazy.Char8 (pack)
import Data.Digest.Pure.MD5 (md5)
import Data.List (find, isPrefixOf)
import Data.Maybe (fromJust)

import AoC (DaySolutions, PartSolution)

nugget :: String -> String -> Int -> Bool
nugget p s i = isPrefixOf p $ show $ md5 $ pack (s ++ show i)

mine :: String -> String -> String
mine prefix s = show $ fromJust $ find (nugget prefix s) [1 ..]

-- exports
part1 :: PartSolution
part1 = mine "00000"

part2 :: PartSolution
part2 = mine "000000"

parts :: DaySolutions
parts = (4, part1, part2)

File: aoe_ruby/lib/2015/day_04.rb
# frozen_string_literal: true

require "digest"

module AoC2015
  class Day04
    def part1(input)
      mine(input, 5)
    end

    def part2(input)
      mine(input, 6)
    end

    private

    def mine(s, zero_count)
      (1..).each do |i|
        return i if Digest::MD5.hexdigest("#{s}#{i}")[0..(zero_count - 1)].chars.uniq == ["0"]
      end
    end
  end
end
